### PR TITLE
Feature/queue rework

### PR DESCRIPTION
**Pull request by PanopticaRising**

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -837,14 +837,14 @@
       }
     },
     "@babel/parser": {
-      "version": "7.12.11",
-      "resolved": "https://registry.npmjs.org/@babel/parser/-/parser-7.12.11.tgz",
-      "integrity": "sha512-N3UxG+uuF4CMYoNj8AhnbAcJF0PiuJ9KHuy1lQmkYsxTer/MAH9UBNHsBoAX/4s6NvlDD047No8mYVGGzLL4hg=="
+      "version": "7.13.9",
```
Comment (red-tt-bear) [inline package-lock.json:840]: Did you accidentally commit the package-lock.json after it was regenerated? There are a lot of changes here

File: src/routes/index.ts [hunk 2]
```diff
@@ -1,79 +1,60 @@
...
 
-    // Respond once the promise to finish is created. The work is done asynchronously above.
-    next(httpResponse.Ok('Working on it!'));
-});
+    const newPriority: PDFPriorityData = { prio: 0, topicId: topic, profUUID: body.professorUUID, firstName: body.firstName };
+
+    // If this is one of the first N PDFs, give elevated priority.
+    if (cheatingInMemoryStorage[topic].pdfPromises.length < configurations.app.highPriorityTabsPerTopic) {
+        logger.debug(`[${topic}] Assigning an initial high priority of 99 to ${body.firstName}`);
+        newPriority.prio = 99;
```
Comment (red-tt-bear) [inline src/routes/index.ts:44]: I thought we were going to change this to Number.Max?

File: src/routes/index.ts [hunk 3]
```diff
@@ -99,11 +80,20 @@ router.get('/', async (_req, _res, next) => {
...
+    }
+    logger.info(`Acquiring a lock to print topic ${topicId}`);
+    const [, release] = await cheatingInMemoryStorage[topicId].lock;
+
     try {
         // Wait for all previous PDF generations for this topic to finish.
         await Promise.allSettled(cheatingInMemoryStorage[topicId].pdfPromises);
     } catch (e) {
         logger.error('Zip was requested before any PDFs were!', e);
+        throw new Error('Zip called too soon.');
```
Comment (red-tt-bear) [inline src/routes/index.ts:96]: Throwing an error with the same message in two different places will cause confusion

**Review by tommy-lettieri — CHANGES_REQUESTED** [review]
package lock should be fine since we are going to need thorough testing anyway
the duplicate error messages should be changed

File: src/routes/index.ts [hunk 4]
```diff
@@ -1,79 +1,60 @@
...
-    .catch(e => logger.error('Route failed to add pdf to zip', e));
+    await cheatingInMemoryStorage[topic].lock;
 
-    // Respond once the promise to finish is created. The work is done asynchronously above.
-    next(httpResponse.Ok('Working on it!'));
-});
+    const newPriority: PDFPriorityData = { prio: 0, topicId: topic, profUUID: body.professorUUID, firstName: body.firstName };
+
+    // If this is one of the first N PDFs, give elevated priority.
+    if (cheatingInMemoryStorage[topic].pdfPromises.length < configurations.app.highPriorityTabsPerTopic) {
```
[the remark this answers is not in the document]
Comment (PanopticaRising) [reply src/routes/index.ts:42]: Items are pushed to this array on line 47 below this, so there will always be a time when this is checked and there are no items in the array.

File: src/routes/index.ts [hunk 5]
```diff
@@ -1,79 +1,60 @@
...
-    cheatingInMemoryStorage[topic].pdfPromises.push(pdfPromise);
+    // If the topic doesn't have a lock yet, acquire one! Then, wait for that lock before processing.
+    if (_.isNil(cheatingInMemoryStorage[topic].lock)) {
+        logger.info(`Creating a lock for topic ${topic}`);
+        cheatingInMemoryStorage[topic].lock = globalTopicSemaphore.acquire();
+    }
 
-    addPDFToZip(cheatingInMemoryStorage[topic].zipObject.archive, pdfPromise, topic)
-    .catch(e => logger.error('Route failed to add pdf to zip', e));
+    await cheatingInMemoryStorage[topic].lock;
```
Comment (red-tt-bear) [inline src/routes/index.ts:37]: What happens if an error occurs after the request responds

i.e. lock times outs

**Review by tommy-lettieri — APPROVED** [review]

**Pull request merged**